flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java

```diff
@@ -188,6 +188,14 @@
 		.defaultValue("")
 		.withDescription("A comma-separated list of tags to apply to the Flink YARN application.");
 
+	/**
+	 * The max number of threads for starting yarn containers in yarn resource manager.
+	 */
+	public static final ConfigOption<Integer> CONTAINER_LAUNCHER_NUM_MAX =
```

Comment:
Do we need to expose this option? I'm wondering how many people will actually set it.

Reply:
I think we need to expose option to make the thread number configurable. We run into many cases in our production where we need to set the thread number to 100 for large scale jobs to meet our SLA for deploying a job cluster. On the other hand, having too many threads consumes significant unnecessary resources for small jobs.

Reply:
I guess I would need to see some benchmarks to see the positive effect of having 100 threads instead of single worker thread. What I could see is if the Yarn site can work on multiple requests in parallel but even then there must be an upper limit to it. Imagine multiple Flink ResourceManager which bombard the Yarn NodeManager with 100s of container requests. This does not feel right.

Reply:
I think the benefit of having many threads starting containers mainly comes from that Flink RM can concurrently start containers from different Yarn NMs. Imagine a job with 1000s of containers on a Yarn cluster of 100s of NMs, ideally Flink RM would be able to concurrently start containers from 100 different NMs, and averagely each NM only have about 10 containers to start. The burden for each NM is affordable even with multiple Flink RMs, because it is not a common case that all the Flink RMs start massive of containers at the same time.

I can try provide some benchmarks. This is not a rare case in our production.

Reply:
Hmm ok, I think the problem arises because we are using the `NMClient` instead of the `NMClientAsync`. Hence the `startContainer` waits for the completion. Wouldn't using the `NMClientAsync` be the much better solution instead of introducing a new `ExecutorService` which makes the `NMClient` asynchronous?